ocaml/xapi/xapi_vdi.ml

```diff
@@ -25,6 +25,30 @@ open Printf
 (**************************************************************************************)
 (* current/allowed operations checking                                                *)
 
+let check_sm_feature_error (op:API.vdi_operations) sm_features sr =
+  let required_sm_feature = Smint.(match op with
+  | `forget
+  | `snapshot
+  | `copy
+  | `scan
+  | `force_unlock
+  | `blocked
+    -> None
+  | `destroy -> Some Vdi_delete
+  | `resize -> Some Vdi_resize
+  | `update -> Some Vdi_update
+  | `resize_online -> Some Vdi_resize_online
+  | `generate_config -> Some Vdi_generate_config
+  | `clone -> Some Vdi_clone
+  | `mirror -> Some Vdi_mirror
+  ) in
+  match required_sm_feature with
+  | None -> None
+  | Some feature ->
+    if Smint.(has_capability feature sm_features)
+    then None
+    else Some (Api_errors.sr_operation_not_supported, [Ref.string_of sr])
+
 (** Checks to see if an operation is valid in this state. Returns Some exception
```

Comment:
I think the practice is that ocamldoc comments aren't duplicated between the mli and ml files, but I may be wrong.